File: Python/위클리챌린지/3주차-퍼즐조각챌린지.py
def rotatePuzzle(puzzle):
    puzzle1 = []
    puzzle2 = []
    puzzle3 = []

    for r,c in puzzle:
        puzzle1.append([c,-r])
        puzzle2.append([-c,r])
        puzzle3.append([-r,-c])
    
    puzzle.sort()
    puzzle1.sort()
    puzzle2.sort()
    puzzle3.sort()
    return [ puzzle,puzzle1,puzzle2,puzzle3]    


def getPuzzle(table):
    N = len(table)
    vis = [[0 for _ in range(N)] for _ in range(N)]
    dr = [0,1,0,-1]
    dc = [1,0,-1,0]
    puzzles = {}
    queue = []
    for i in range(N):
        for j in range(N):
            if table[i][j] == 1 and vis[i][j] == 0:
                queue.append([i,j])
                vis[i][j] = 1
                puzzle = []
                while(queue):
                    r,c = queue.pop();
                    puzzle.append([r-i,c-j])
                    for k in range(4):
                        cr = r+dr[k]
                        cc = c+dc[k]
                        if cr < 0 or cr >= N or cc < 0 or cc >=N : continue
                        if vis[cr][cc] == 1: continue
                        if table[cr][cc] == 0: continue
                        vis[cr][cc] = 1
                        queue.append([cr,cc])
                puzzles[len(puzzles)] = rotatePuzzle(puzzle)
    return puzzles

def fixHole(table,puzzles):
    fixedHole = 0
    N = len(table)
    used = [0 for _ in range(len(puzzles))]
    dr = [0,1,0,-1]
    dc = [1,0,-1,0]
    queue = []
    for i in range(N):
        for j in range(N):
            if table[i][j] == 0:
                vis = [[0 for _ in range(N)] for _ in range(N)]
                queue.append([i,j])
                vis[i][j] = 1
                hole = []
                while(queue):
                    r,c = queue.pop();
                    hole.append([r-i,c-j])
                    for k in range(4):
                        cr = r+dr[k]
                        cc = c+dc[k]
                        if cr < 0 or cr >= N or cc < 0 or cc >=N : continue
                        if vis[cr][cc] == 1: continue
                        if table[cr][cc] == 1: continue
                        vis[cr][cc] = 1
                        queue.append([cr,cc])
                fixed = False
                hole.sort()
                print("hole ")
                print(hole)
                for k in range(len(puzzles)):
                    if fixed: break
                    if used[k] == 1: continue
                    if len(hole) != len(puzzles[k][0]): continue
                    # print("puzzle")
                    for puzzle in puzzles[k]:
                        # print(puzzle)
                        if hole == puzzle:
                            fixedHole += len(hole)
                            fixed = True
                            used[k] = 1
                            print(hole, puzzle)
                            for x,y in hole:
                                table[i+x][j+y] = 1
                            break
    return fixedHole


def solution(game_board, table):
    puzzles = getPuzzle(table)
    return fixHole(game_board,puzzles);



# print(solution([[1,1,0,0,1,0],[0,0,1,0,1,0],[0,1,1,0,0,1],[1,1,0,1,1,1],[1,0,0,0,1,0],[0,1,1,1,0,0]],[[1,0,0,1,1,0],[1,0,1,0,1,0],[0,1,1,0,1,1],[0,0,1,0,0,0],[1,1,0,1,1,0],[0,1,0,0,0,0]]))
# print(solution([[0,0,0],[1,1,0],[1,1,1]],[[1,1,1],[1,0,0],[0,0,0]]))
print(solution([[0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0], [1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0], [0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0], [1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1], [0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0], [0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1], [0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0], [0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0], [1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0], [0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0], [0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1], [0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0]],[[1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1], [1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1], [1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0], [0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0], [1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0], [1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0], [1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1], [1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1], [0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1], [1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1], [1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1], [1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1]]))
#54
# print(solution([[1,0,0],[1,0,0],[1,0,0]], [[0,1,1],[0,1,1],[0,1,1]])) #6
# print(solution([[0,0,0],[0,0,0],[0,0,0]], [[1,1,1],[1,1,1],[1,1,1]])) #9

File: Javascript/프로그래머스/위클리챌린지/3주차-퍼즐조각챌린지.js
function rotatePuzzle(puzzle) {
    let puzzle90 = [];
    let puzzle180 = [];
    let puzzle270 = [];

    for (let [r, c] of puzzle) {
        puzzle90.push([c, -r]);
        puzzle180.push([-c, r]);
        puzzle270.push([-r, -c]);
    }
    puzzle.sort((a, b) => (a[0] - b[0] === 0 ? a[1] - b[1] : a[0] - b[0]));
    puzzle90.sort((a, b) => (a[0] - b[0] === 0 ? a[1] - b[1] : a[0] - b[0]));
    puzzle180.sort((a, b) => (a[0] - b[0] === 0 ? a[1] - b[1] : a[0] - b[0]));
    puzzle270.sort((a, b) => (a[0] - b[0] === 0 ? a[1] - b[1] : a[0] - b[0]));

    return [puzzle, puzzle90, puzzle180, puzzle270];
}

function getPuzzles(table) {
    const N = table.length;
    const dr = [0, 1, 0, -1];
    const dc = [1, 0, -1, 0];
    let puzzles = new Map();
    let stack = [];

    for (let i of Array(N).keys()) {
        for (let j of Array(N).keys()) {
            if (table[i][j] === 1) {
                stack.push([i, j]);
                table[i][j] = 0;
                let puzzle = [];
                while (stack.length) {
                    let [r, c] = stack.pop();
                    puzzle.push([r - i, c - j]);
                    for (let k of Array(4).keys()) {
                        let cr = r + dr[k];
                        let cc = c + dc[k];
                        if (cr < 0 || cr >= N || cc < 0 || cc >= N) continue;
                        if (table[cr][cc] === 0) continue;
                        table[cr][cc] = 0;
                        stack.push([cr, cc]);
                    }
                }
                puzzles[Object.keys(puzzles).length] = rotatePuzzle(puzzle);
            }
        }
    }
    return puzzles;
}

function fixHoles(table, puzzles) {
    const N = table.length;
    const dr = [0, 1, 0, -1];
    const dc = [1, 0, -1, 0];
    let fixedHole = 0;
    let used = new Array(Object.keys(puzzles).length).fill(0);
    let stack = [];
    for (let i of Array(N).keys()) {
        for (let j of Array(N).keys()) {
            if (table[i][j] === 0) {
                let vis = Array.from(Array(N), () => Array(N).fill(0));
                stack.push([i, j]);
                vis[i][j] = 1;
                let hole = [];
                while (stack.length) {
                    let [r, c] = stack.pop();
                    hole.push([r - i, c - j]);
                    for (let k of Array(4).keys()) {
                        let cr = r + dr[k];
                        let cc = c + dc[k];
                        if (cr < 0 || cr >= N || cc < 0 || cc >= N) continue;
                        if (vis[cr][cc] === 1) continue;
                        if (table[cr][cc] === 1) continue;
                        vis[cr][cc] = 1;
                        stack.push([cr, cc]);
                    }
                }
                let fixed = false;
                hole.sort((a, b) => (a[0] - b[0] === 0 ? a[1] - b[1] : a[0] - b[0]));
                for (let k of Array(Object.keys(puzzles).length).keys()) {
                    if (fixed) break;
                    if (used[k] === 1) continue;
                    if (hole.length !== puzzles[k][0].length) continue;
                    for (let puzzle of puzzles[k]) {
                        if (JSON.stringify(hole) === JSON.stringify(puzzle)) {
                            fixedHole += hole.length;
                            fixed = true;
                            used[k] = 1;
                            for (let [x, y] of hole) {
                                table[i + x][j + y] = 1;
                            }
                            break;
                        }
                    }
                }
            }
        }
    }
    return fixedHole;
}

function solution(game_board, table) {
    const puzzles = getPuzzles(table);
    return fixHoles(game_board, puzzles);
}

console.log(
    solution(
        [
            [1, 1, 0, 0, 1, 0],
            [0, 0, 1, 0, 1, 0],
            [0, 1, 1, 0, 0, 1],
            [1, 1, 0, 1, 1, 1],
            [1, 0, 0, 0, 1, 0],
            [0, 1, 1, 1, 0, 0],
        ],
        [
            [1, 0, 0, 1, 1, 0],
            [1, 0, 1, 0, 1, 0],
            [0, 1, 1, 0, 1, 1],
            [0, 0, 1, 0, 0, 0],
            [1, 1, 0, 1, 1, 0],
            [0, 1, 0, 0, 0, 0],
        ]
    )
);
